polaris-core/src/main/java/org/apache/polaris/core/catalog/PageToken.java

```diff
@@ -0,0 +1,145 @@
+ * response, that means there is no more data to read.
+ */
+public class PageToken {
+
+  public final int offset;
+  public final int pageSize;
+
+  private static final String TOKEN_PREFIX = "polaris";
+  private static final int TOKEN_START = 0;
+  private static final int DEFAULT_PAGE_SIZE = 1000;
+
+  public static PageToken DONE = null;
+
+  public PageToken(int offset, int pageSize) {
+    this.offset = offset;
+    this.pageSize = pageSize;
+  }
+
+  /** Construct a PageToken from a plain limit */
+  public static PageToken fromLimit(int limit) {
+    return new PageToken(TOKEN_START, limit);
+  }
+
+  /** Construct a PageToken to read everything */
+  public static PageToken readEverything() {
+    return new PageToken(TOKEN_START, Integer.MAX_VALUE);
+  }
+
+  /** Deserialize a token string into a PageToken object */
+  public static PageToken fromString(String tokenString) {
+    if (tokenString == null) {
+      return PageToken.readEverything();
+    } else if (tokenString.isEmpty()) {
+      return PageToken.fromLimit(DEFAULT_PAGE_SIZE);
+    } else {
+      try {
+        String decoded =
+            new String(Base64.getDecoder().decode(tokenString), StandardCharsets.UTF_8);
+        String[] parts = decoded.split(":");
+
+        if (parts.length != 4 || !parts[0].equals(TOKEN_PREFIX)) {
+          throw new IllegalArgumentException("Invalid token format in token: " + tokenString);
+        }
+
+        int offset = Integer.parseInt(parts[1]);
+        int pageSize = Integer.parseInt(parts[2]);
+        int checksum = Integer.parseInt(parts[3]);
+        PageToken token = new PageToken(offset, pageSize);
+
+        if (token.hashCode() != checksum) {
+          throw new IllegalArgumentException("Invalid checksum for token: " + tokenString);
+        } else {
+          return token;
+        }
+      } catch (Exception e) {
+        throw new IllegalArgumentException("Failed to decode page token: " + tokenString, e);
+      }
+    }
+  }
+
+  /**
+   * Builds a new page token to reflect new data that's been read. If the amount of data read is
+   * less than the pageSize, this will return `PageToken.DONE` (done)
+   */
+  public PageToken updated(List<?> newData) {
+    if (newData == null || newData.isEmpty() || newData.size() < pageSize) {
+      return PageToken.DONE;
+    } else {
+      return new PageToken(offset + newData.size(), pageSize);
+    }
+  }
+
+  /**
+   * Builds a `PolarisPage<T>` from a `List<T>`. The `PageToken` attached to the new
+   * `PolarisPage<T>` is the same as the result of calling `updated(data)` on this `PageToken`.
+   */
+  public <T> PolarisPage<T> buildNextPage(List<T> data) {
+    return new PolarisPage<T>(this.updated(data), data);
+  }
+
+  /**
+   * Return a new PageToken with an updated pageSize. If the pageSize provided is null, the existing
+   * pageSize will be preserved.
+   */
+  public PageToken withPageSize(Integer pageSize) {
+    if (pageSize == null) {
```

Comment:
Do we need any validation for pageSize <= 0?

Reply:
The spec already sets a minimum of 1 for the pageSize [here](https://github.com/apache/iceberg/blob/ab2c6f889d07eeee51a1f58605be248e9330d91b/open-api/rest-catalog-open-api.yaml#L1573). The check doesn't go in this class, but at the API level.